### PR TITLE
Add Virtual Numpad support

### LINKED ISSUE #8719: Editor: Add on-screen keypad for all number based inputs in the editor
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

On touch-screen devices, setting army size / resources rewards is troublesome, as there is no way to enter the values directly and only up/down arrows are available. Hence entering higher values takes a lot of time.

Also currently it is not possible to set negative values of event reward.

Affects all touch-screen devices.

### Describe the possible solution

Some kind of on-screen keypad (similar to already implemented on-screen keyboard, but only with numbers and +/- sign for event rewards) could be introduced for touch screen devices.

### Additional info

version 1.1.0 on Android 5.1.1

### PR DESCRIPTION
**Pull request by ihhub**
close #8719

Screenshot 2024-11-20 at 9 12 06 PM

### CODE REVIEW

File: src/fheroes2/gui/ui_keyboard.cpp [hunk 1]
```diff
@@ -55,6 +55,7 @@ namespace
     const int32_t defaultSpecialButtonWidth{ 54 };
     const int32_t spacebarButtonWidth{ 175 };
     const int32_t defaultWindowWidth{ 520 };
+    const int32_t numbericWindowWidth{ 320 };
```
Comment (ihhub) [inline src/fheroes2/gui/ui_keyboard.cpp:58]: This is spelled incorrectly.

File: src/fheroes2/gui/ui_keyboard.cpp [hunk 2]
```diff
@@ -493,6 +528,29 @@ namespace
...
+            } );
+
+            lastButtonRow.emplace_back( "0", getDefaultButtonWidth( fheroes2::SupportedLanguage::English ), isEvilInterface, []( KeyboardRenderer & renderer ) {
+                renderer.insertCharacter( '0' );
+                return DialogAction::AddLetter;
+            } );
+
+            lastButtonRow.emplace_back( "\x7F", getDefaultButtonWidth( fheroes2::SupportedLanguage::English ), isEvilInterface,
+                                        []( const KeyboardRenderer & ) { return DialogAction::ChangeLanguage; } );
+            lastButtonRow.back().button.hide();
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/gui/ui_keyboard.cpp:552]: To center the rest of button.

File: src/fheroes2/gui/ui_keyboard.cpp [hunk 3]
```diff
@@ -784,4 +843,36 @@ namespace fheroes2
...
+            case DialogAction::Backspace:
+            case DialogAction::LowerCase:
+            case DialogAction::UpperCase:
+            case DialogAction::AlphaNumeric:
+            case DialogAction::ChangeLanguage:
+                // These actions must not be processed here!
+                assert( 0 );
+                break;
+            case DialogAction::Close:
+                output = std::stoi( outputString );
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/gui/ui_keyboard.cpp:869]: It is a good point. We should limit the number of digits.

[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/gui/ui_keyboard.cpp:869]: Or, as a variant, we can catch this exception (just this one, there is also `std::invalid_argument`, but it should never be normally raised if the logic is fine) and handle it by showing some error dialog maybe (and not change the number in the corresponding input).

File: src/fheroes2/gui/ui_keyboard.cpp [hunk 4]
```diff
@@ -730,18 +792,16 @@ namespace fheroes2
 {
     void openVirtualKeyboard( std::string & output )
     {
-        const std::vector<SupportedLanguage> supportedLanguages = getSupportedLanguages();
```
Comment (oleg-derevenetz) [inline src/fheroes2/gui/ui_keyboard.cpp:733]: What was the purpose of this line? The `supportedLanguages` variable was unused from the very beginning.

**Review by oleg-derevenetz — COMMENTED** [review]
Hi @ihhub I just made a few changes regarding the sanity checking improvements that should fix [this](https://github.com/ihhub/fheroes2/pull/9287#discussion_r1852681411) my comment. Also I renamed the `numbericWindowWidth` constant, which should fix also [this](https://github.com/ihhub/fheroes2/pull/9287#discussion_r1851537797) comment.

File: src/fheroes2/gui/ui_keyboard.cpp (hunk 4 above)
Comment (ihhub) [reply src/fheroes2/gui/ui_keyboard.cpp:733]: I think it's a leftover from the old code.

File: src/fheroes2/gui/ui_keyboard.cpp [hunk 5]
```diff
@@ -784,4 +844,59 @@ namespace fheroes2
...
+                const auto handleError = [&action]() {
+                    showStandardTextMessage( _( "Warning" ), _( "The entered value is invalid." ), Dialog::OK );
+
+                    action = DialogAction::DoNothing;
+                };
+
+                try {
+                    const int32_t intValue = std::stoi( strValue );
+                    if ( intValue < minValue || intValue > maxValue ) {
+                        handleError();
```
Comment (Districh-ru) [inline src/fheroes2/gui/ui_keyboard.cpp:878]: Should we show the limits values in the error message?
Currently it is not shown and players will spend some annoying time to find this limit.

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi, @ihhub,
After the `handleError` dialog is shown th **OK** button stays displayed as pressed:
изображение

Also I left one suggestion. Could you take a look when you have time.

**Comment by Districh-ru** [comment]
Hi, @ihhub, I've fixed the found issue and implemented a suggestion to show value limits when the entered value is out of bounds.

**Review by Districh-ru — APPROVED** [review]

**Review by oleg-derevenetz — APPROVED** [review]

**Pull request merged**